tests/middleware_test.py

```diff
@@ -0,0 +1,222 @@
+                base_url=f"http://{mcp.settings.host}:{mcp.settings.port}",
+            )
+            g2_client = httpx.AsyncClient(
+                base_url=f"http://{mcp.settings.host}:{mcp.settings.port}"
+            )
+
+            # Monkeypatch the MCP server to use our middleware.
+            mcp.run_sse_async = MethodType(run_sse_async_with_middleware, mcp)
+
+            async with anyio.create_task_group() as tg:
+                tg.start_soon(mcp.run_sse_async, name="mcp")
+                # Wait for the server to start.
+                await asyncio.sleep(0.1)
+
+                # Send SSE requests to the MCP server, one for each Grafana server.
+                # We can access tool call results over the SSE stream.
+                async with (
+                    aconnect_sse(
+                        g1_client,
+                        "GET",
+                        "/sse",
+                        headers={
+                            "X-Grafana-Url": f"http://{g1.host}:{g1.port}",
+                            "X-Grafana-Api-Key": "abcd123",
+                        },
+                    ) as g1_source,
+                    aconnect_sse(
+                        g2_client,
+                        "GET",
+                        "/sse",
+                        headers={
+                            "X-Grafana-Url": f"http://{g2.host}:{g2.port}",
+                            "X-Grafana-Api-Key": "efgh456",
+                        },
+                    ) as g2_source,
+                ):
+                    g1_iter = g1_source.aiter_sse()
+                    g2_iter = g2_source.aiter_sse()
+                    # The URL to use is in the first SSE message.
+                    g1_url = (await g1_iter.__anext__()).data
+                    g2_url = (await g2_iter.__anext__()).data
+
+                    # The MCP protocol requires us to send an initialize request
+                    # before we can send any other requests.
+                    await initialize(g1_client, g1_url, g1_iter)
+                    await initialize(g2_client, g2_url, g2_iter)
+
+                    # Send a tool call request using the first URL.
+                    await g1_client.post(
+                        g1_url,
+                        json={
+                            "jsonrpc": "2.0",
+                            "id": 2,
+                            "method": "tools/call",
+                            "params": {"name": "list_datasources"},
+                        },
+                    )
+                    result = await jsonrpc_result(g1_iter)
+                    # This must have come from the first Grafana server.
+                    assert json.loads(result.content[0].text) == json.dumps(  # type: ignore
+                        [{"id": 1}], indent=4
+                    )
+
+                    # Send a tool call request using the second URL.
+                    await g2_client.post(
```

Comment:
Wondering if running the two requests in parallel will make the test a bit more convincing?